File: src/rust/bitbox02-rust/src/hww/api/restore.rs
// Copyright 2021 Shift Crypto AG
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use super::Error;
use crate::pb;

use pb::response::Response;

use crate::hal::Ui;
use crate::workflow::{confirm, mnemonic, password, unlock};

pub async fn from_file(
    hal: &mut impl crate::hal::Hal,
    request: &pb::RestoreBackupRequest,
) -> Result<Response, Error> {
    // This is a separate screen because 'Restore backup?' does not fit in the title field.
    hal.ui()
        .confirm(&confirm::Params {
            body: "Restore backup?",
            accept_is_nextarrow: true,
            ..Default::default()
        })
        .await?;

    let (data, metadata) = match crate::backup::load(hal, &request.id).await {
        Ok(d) => d,
        Err(_) => {
            hal.ui().status("Could not\nrestore backup", false).await;
            return Err(Error::Generic);
        }
    };

    hal.ui()
        .confirm(&confirm::Params {
            body: &format!("Name: {}. ID: {}", &metadata.name, &request.id),
            scrollable: true,
            accept_is_nextarrow: true,
            ..Default::default()
        })
        .await?;

    #[cfg(feature = "app-u2f")]
    {
        let datetime_string =
            bitbox02::format_datetime(request.timestamp, request.timezone_offset, false)
                .map_err(|_| Error::InvalidInput)?;
        let params = confirm::Params {
            title: "Is now?",
            body: &datetime_string,
            accept_is_nextarrow: true,
            ..Default::default()
        };
        hal.ui().confirm(&params).await?;
    }

    let password = password::enter_twice(hal).await?;
    let seed = data.get_seed();
    if let Err(err) = crate::keystore::encrypt_and_store_seed(hal, seed, &password) {
        hal.ui()
            .status(&format!("Could not\nrestore backup\n{:?}", err), false)
            .await;
        return Err(Error::Generic);
    }

    // Ignore error here. Missing birthdate should not abort an otherwise successful restore.
    let _ = bitbox02::memory::set_seed_birthdate(data.0.birthdate);

    #[cfg(feature = "app-u2f")]
    {
        // Ignore error - the U2f counter not being set can lead to problems with U2F, but it should
        // not fail the recovery, so the user can access their coins.
        let _ = bitbox02::securechip::u2f_counter_set(request.timestamp);
    }

    bitbox02::memory::set_initialized().or(Err(Error::Memory))?;

    // Ignore non-critical error.
    let _ = bitbox02::memory::set_device_name(&metadata.name);

    unlock::unlock_bip39(hal, seed).await;
    Ok(Response::Success(pb::Success {}))
}

pub async fn from_mnemonic(
    hal: &mut impl crate::hal::Hal,
    #[cfg_attr(not(feature = "app-u2f"), allow(unused_variables))]
    &pb::RestoreFromMnemonicRequest {
        timestamp,
        timezone_offset,
    }: &pb::RestoreFromMnemonicRequest,
) -> Result<Response, Error> {
    #[cfg(feature = "app-u2f")]
    {
        let datetime_string = bitbox02::format_datetime(timestamp, timezone_offset, false)
            .map_err(|_| Error::InvalidInput)?;
        hal.ui()
            .confirm(&confirm::Params {
                title: "Is now?",
                body: &datetime_string,
                accept_is_nextarrow: true,
                ..Default::default()
            })
            .await?;
    }

    let mnemonic = mnemonic::get(hal).await?;
    let seed = match crate::bip39::mnemonic_to_seed(&mnemonic) {
        Ok(seed) => seed,
        Err(()) => {
            hal.ui().status("Recovery words\ninvalid", false).await;
            return Err(Error::Generic);
        }
    };
    hal.ui().status("Recovery words\nvalid", true).await;

    // If entering password fails (repeat password does not match the first), we don't want to abort
    // the process immediately. We break out only if the user confirms.
    let password = loop {
        match password::enter_twice(hal).await {
            Err(password::EnterTwiceError::DoNotMatch) => {
                hal.ui()
                    .confirm(&confirm::Params {
                        title: "",
                        body: "Passwords\ndo not match.\nTry again?",
                        ..Default::default()
                    })
                    .await?;
            }
            Err(err @ password::EnterTwiceError::EnterError(_)) => return Err(err.into()),
            Ok(password) => break password,
        }
    };

    if let Err(err) = crate::keystore::encrypt_and_store_seed(hal, &seed, &password) {
        hal.ui()
            .status(&format!("Could not\nrestore backup\n{:?}", err), false)
            .await;
        return Err(Error::Generic);
    };

    #[cfg(feature = "app-u2f")]
    {
        // Ignore error - the U2f counter not being set can lead to problems with U2F, but it should
        // not fail the recovery, so the user can access their coins.
        let _ = bitbox02::securechip::u2f_counter_set(timestamp);
    }

    bitbox02::memory::set_initialized().or(Err(Error::Memory))?;

    unlock::unlock_bip39(hal, &seed).await;
    Ok(Response::Success(pb::Success {}))
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::hal::testing::TestingHal;
    use bitbox02::testing::mock_memory;
    use bitbox02::{keystore, memory};
    use util::bb02_async::block_on;

    use alloc::boxed::Box;

    #[test]
    fn test_from_mnemonic() {
        mock_memory();
        crate::keystore::lock();
        let mut counter = 0u32;
        let mut mock_hal = TestingHal::new();
        mock_hal.ui.set_enter_string(Box::new(|params| {
            counter += 1;
            match counter {
                1 => assert_eq!(params.title, "Set password"),
                2 => assert_eq!(params.title, "Repeat password"),
                _ => panic!("too many user inputs"),
            }
            Ok("password".into())
        }));

        mock_hal.securechip.event_counter_reset();
        assert_eq!(
            block_on(from_mnemonic(
                &mut mock_hal,
                &pb::RestoreFromMnemonicRequest {
                    timestamp: 0,
                    timezone_offset: 0,
                }
            )),
            Ok(Response::Success(pb::Success {}))
        );
        assert_eq!(mock_hal.securechip.get_event_counter(), 8);
        drop(mock_hal); // to remove mutable borrow of counter
        assert_eq!(counter, 2);
        assert!(!crate::keystore::is_locked());
        assert!(memory::is_initialized());
        // Seed of hardcoded phrase used in unit tests:
        // boring mistake dish oyster truth pigeon viable emerge sort crash wire portion cannon couple enact box walk height pull today solid off enable tide
        assert_eq!(
            hex::encode(crate::keystore::copy_seed().unwrap()),
            "19f1bcfccf3e9d497cd245cf864ff0d42216625258d4f68d56b571aceb329257"
        );
        assert_eq!(
            hex::encode(crate::keystore::copy_bip39_seed().unwrap()),
            "257724bccc8858cfe565b456b01263a4a6a45184fab4531f5c199649207a74e74c399a01d4f957258c05cee818369b31404c884a4b7a29ff6886bae6700fb56a"
        );
    }
}

File: src/backup/restore.c
// Copyright 2019 Shift Cryptosecurity AG
// Copyright 2020 Shift Crypto AG
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "restore.h"
#include "backup_common.h"

#include <stdio.h>
#include <string.h>

#include <keystore.h>
#include <memory/memory.h>
#include <sd.h>
#include <util.h>

#include <pb_decode.h>
#include <wally_crypto.h>

/**
 * Data used during decode.
 */
typedef struct decode_data {
    BackupData* backup_data;
    BackupMode* mode;
} decode_data_t;

typedef struct {
    const char* filename;
    uint8_t content[SD_MAX_FILE_SIZE];
    uint32_t file_length;
    bool loaded;
} broken_backup_t;

typedef struct {
    uint8_t count;
    broken_backup_t backups[3];
} broken_backups_t;

typedef struct {
    uint8_t content[SD_MAX_FILE_SIZE];
    uint32_t file_length;
    Backup* backup;
    BackupData* backup_data;
} good_backup_t;

static bool _decode_backup_data(pb_istream_t* istream, const pb_field_t* field, void** arg)
{
    (void)field;
    decode_data_t* decode_data = (decode_data_t*)*arg;
    if (*(decode_data->mode) != BackupMode_PLAINTEXT) {
        return false;
    }
    if (!pb_decode(istream, BackupData_fields, decode_data->backup_data)) {
        return false;
    }
    return true;
}

restore_error_t restore_from_buffer(
    uint8_t* input,
    uint32_t length,
    Backup* backup,
    BackupData* backup_data)
{
    decode_data_t decode_data;
    BackupContent* backup_content = &backup->backup_v1.content;
    decode_data.mode = &backup_content->metadata.mode;
    decode_data.backup_data = backup_data;
    backup->backup_v1.content.data.arg = &decode_data;
    backup->backup_v1.content.data.funcs.decode = &_decode_backup_data;

    pb_istream_t in_stream = pb_istream_from_buffer(input, length);
    bool status = pb_decode(&in_stream, Backup_fields, backup);
    if (!status) {
        return RESTORE_ERR_DECODE;
    }
    return RESTORE_OK;
}
